### PR TITLE
Support configuring environments via .slashdeploy.yml

### PR DESCRIPTION
**Pull request by ejholmes**
Closes https://github.com/remind101/slashdeploy/issues/94
Fixes https://github.com/remind101/slashdeploy/issues/50

Since the move to a GitHub Integration, it makes it easy to support configuring a repositories known environments (and continuous delivery settings), via a file in the root of the repository, similar to something like `.circleci` or `.travis.yml`.

With this change, you can now add a `.slashdeploy.yml` to the root of a repositories default branch, and SlashDeploy will use that as it's source for determining what environments can be deployed to.

Here's a simple example that configures two environments, one in which we continously deliver the "master" branch to:

```yaml
---
production:
  
  # Allows us to reference this using "production" or a short alias, "prod".
  aliases:
    - prod

  # Enable continuous delivery for this environment.
  continuous_delivery:
    # You can specify any git "ref" (tag, branch, etc) to auto deploy when
    # pushed to.
    ref: refs/heads/master

    # If you want to only deploy when a set of GitHub commit statuses are
    # passing, you can configure those here. If none are provided, SlashDeploy
    # will deploy when new commits are pushed to the git ref.
    required_contexts:
      - ci/circleci

staging:
  aliases:
    - stage
```

I think this is a big improvement over what we had before (basically nothing), and better than having some kind of UI to do the configuration.

### CODE REVIEW

File: app/views/documentation/index.html.erb [hunk 1]
```diff
@@ -3,7 +3,7 @@
     <ul class="nav nav-pills nav-stacked">
       <li role="presentation"><a href="#installation">Installation</a></li>
       <li role="presentation"><a href="#slash-command">Slash Command</a></li>
-      <li role="presentation"><a href="#auto-deployment">Auto Deployment</a></li>
+      <li role="presentation"><a href="#continuous-delivery">Continuous Delivery</a></li>
```
Comment (russellballestrini) [inline app/views/documentation/index.html.erb]: Nice, I like this more.

File: app/views/documentation/index.html.erb [hunk 2]
```diff
@@ -42,8 +42,29 @@
...
+    # You can specify any git "ref" (tag, branch, etc) to auto deploy when
+    # pushed to.
+    ref: refs/heads/master
+
+    # If you want to only deploy when a set of GitHub commit statuses are
+    # passing, you can configure those here. If none are provided, SlashDeploy
+    # will deploy when new commits are pushed to the git ref.
+    required_contexts:
+      - ci/circleci</pre>
+    </p>
```
Comment (russellballestrini) [inline app/views/documentation/index.html.erb]: Do you think it's worth talking about the SlashDeploy Github Integration `Single File` setting?

File: lib/slashdeploy/config.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  #         - prod
+  #       continuous_delivery:
+  #         ref: refs/heads/master
+  #         required_contexts:
+  #           - ci/circleci
+  #     staging:
+  #       aliases:
+  #         - stage
+  class Config
+    include Virtus.model
```
Comment (russellballestrini) [inline lib/slashdeploy/config.rb]: Just for my own understanding, why is this included in each method individually instead of just "at the top"?

File: app/models/environment.rb [hunk 4]
```diff
@@ -96,6 +128,23 @@ def installation
     repository.installation
   end
 
+  def check_config!
+    fail("This repository is configured from .slashdeploy.yaml") if repository && repository.config?
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply app/models/environment.rb]: Yeah, this will throw an exception. It's mostly here, since calling these methods on a repo that's configured with a `.slashdeploy.yml` would be a programmer error. This should never actually get hit, but I wan't to fail loudly if it does.

File: lib/slashdeploy/config.rb (hunk 3 above)
Comment (ejholmes) [reply lib/slashdeploy/config.rb]: If a class is nested under another class, it doesn't inherit anything from the class above it. Generally when you nest classes like this, you're only doing it to namespace the class (e.g. `SlashDeploy::Config::Environment`).

File: app/models/environment.rb [hunk 5]
```diff
@@ -96,6 +128,23 @@ def installation
...
+    config.present?
+  end
+
+  def continuous_delivery_config?
+    config? && config.continuous_delivery.present?
+  end
+
+  def config
+    return unless repository && repository.config?
+    repository.config.environments[name]
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply app/models/environment.rb]: I guess this is a style thing. There's two ways people generally code conditions:

**if/else**

```ruby
if thing
  this
else
  that
end
```

**early return**

```ruby
return this if thing
that
```

In general, I prefer that latter (early return) just because it's less cognitive load if a method is large (if it returns, you can just stop reading). In this case, the early return is really a "guard".

File: app/models/environment.rb [hunk 6]
```diff
@@ -49,14 +49,36 @@ def locked?
     active_lock.present?
   end
 
+  # Returns true if the provided name, matches the canonical environment name,
+  # or one of it's aliases.
+  def match_name?(name)
+    return true if self.name == name
+    return true if config.aliases.include?(name)
+    false
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply app/models/environment.rb]: Only reason is to make the method consistent with the type it returns (true/false). If I leave it out, it'd return nil. In Ruby, that's technically fine, because both `nil` and `false` are considered "falsey".

File: app/views/documentation/index.html.erb (hunk 2 above)
Comment (ejholmes) [reply app/views/documentation/index.html.erb]: This doc is more about usage docs. But I should definitely add this to the setup docs, if anybody wants to run SlashDeploy themselves.

File: app/models/environment.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (russellballestrini) [reply app/models/environment.rb]: I guess what I'm trying to say is, explicit `return false` instead of implicit `false` if the block already has other return statement conditions.

File: app/models/environment.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (russellballestrini) [reply app/models/environment.rb]: Cool.

File: app/views/documentation/index.html.erb (hunk 2 above)
Comment (russellballestrini) [reply app/views/documentation/index.html.erb]: Agreed. I still think that is a decent goal, it's pretty cool.

Another reason to document it in the setup docs is that if SlashDeploy integration gets uninstalled by accident or something.
